### PR TITLE
fileter list bucket objects based on ranger policies

### PR DESCRIPTION
**Pull request by kr7ysztof**
For the recursive list bucket - currently there is checked only the root resource in ranger so it is posible to see all subdirs. The fix checks in ranger all subdirs.

### CODE REVIEW

File: src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala [hunk 1]
```diff
@@ -50,4 +64,101 @@ trait RequestHandlerS3 extends LazyLogging with RadosGatewayHandler {
...
+   * for list objects in bucket we need filter response when --recursive is set
+   * --recursive means that there is a delimiter param in request query and the access type is READ
+   * @param request
+   * @param userSTS
+   * @param s3request
+   * @param response
+   * @return for recursive request it returns filtered response for others the original response
+   */
+  protected[this] def filterResponse(request: HttpRequest, userSTS: User, s3request: S3Request, response: HttpResponse): HttpResponse = {
+    println(s3request)
```
Comment (nielsdenissen) [inline src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala]: remnant of dev I guess ;)

File: src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala [hunk 2]
```diff
@@ -1,14 +1,23 @@
 package com.ing.wbaa.airlock.proxy.handler
```
Comment (nielsdenissen) [inline src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala]: I would propose to move the functionality of extracting the request and checking authorization as we do here to a separate class. We can separate concerns a bit more that way and keep this class responsible only for actually handling requests with S3 (not also checking ranger, unfolding request etc.)

File: src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala [hunk 3]
```diff
@@ -49,5 +63,102 @@ trait RequestHandlerS3 extends LazyLogging with RadosGatewayHandler {
...
+              elementResult(allContentsElements, isContentsTag, element)
+            case element: EndElement if element.localName == "Key" =>
+              isKeyTag = false
+              elementResult(allContentsElements, isContentsTag, element)
+            case element: TextEvent =>
+              if (isKeyTag) keyTagValue.append(element.text)
+              elementResult(allContentsElements, isContentsTag, element)
+            case s =>
+              elementResult(allContentsElements, isContentsTag, s)
+          }
```
Comment (nielsdenissen) [inline src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala]: Could you add a brief comment on what each case means or a java doc on top of the function? It's quite a large function and not immediately clear to me what it does/how it works

File: src/main/scala/com/ing/wbaa/airlock/proxy/provider/MessageProviderKafka.scala [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
 trait MessageProviderKafka extends EventProducer with AWSMessageEventJsonSupport {
 
   private def incompleteData: Future[Nothing] = Future.failed(new Exception("Cannot send event to kafka, not enough input data"))
 
-  def emitEvent(s3Request: S3Request, method: HttpMethod, principalId: String, clientIPAddress: RemoteAddress): Future[Done] =
+  def emitEvent(s3Request: S3Request, method: HttpMethod, principalId: String): Future[Done] =
     method match {
       case HttpMethods.POST | HttpMethods.PUT =>
-        prepareAWSMessage(s3Request, method, principalId, clientIPAddress, s3ObjectCreated(method.value))
+        prepareAWSMessage(s3Request, method, principalId, s3Request.clientIPAddress.get, s3ObjectCreated(method.value))
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/airlock/proxy/provider/MessageProviderKafka.scala]: changed

File: src/main/scala/com/ing/wbaa/airlock/proxy/provider/MessageProviderKafka.scala [hunk 5]
```diff
@@ -1,28 +1,28 @@
...
           .map { case jse =>
             sendSingleMessage(jse.toString(), kafkaSettings.createEventsTopic)
           }
           .getOrElse(incompleteData)
 
       case HttpMethods.DELETE =>
-        prepareAWSMessage(s3Request, method, principalId, clientIPAddress, s3ObjectRemoved(method.value))
+        prepareAWSMessage(s3Request, method, principalId, s3Request.clientIPAddress.get, s3ObjectRemoved(method.value))
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/airlock/proxy/provider/MessageProviderKafka.scala]: changed

File: src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala (hunk 1 above)
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala]: removed

File: src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala (hunk 2 above)
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala]: done

File: src/test/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3Spec.scala [hunk 6]
```diff
@@ -28,15 +32,51 @@ class RequestHandlerS3Spec extends AsyncWordSpec with DiagrammedAssertions with
...
   }
+
+  val listBucketXmlResponse: String = scala.io.Source.fromResource("listBucket.xml").mkString.stripMargin.trim
+
+  val adminUser = User(UserRawJson("admin", Set.empty[String], "a", "s"))
+  val user1 = User(UserRawJson("user1", Set.empty[String], "a", "s"))
+  val s3Request = S3Request(AwsRequestCredential(AwsAccessKey(""), None), Uri.Path("/demobucket/user"), HttpMethods.GET)
+  val data: Source[ByteString, NotUsed] = Source.single(ByteString.fromString(listBucketXmlResponse))
+
+  "List bucket object response" should {
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/test/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3Spec.scala]: done

File: src/main/scala/com/ing/wbaa/airlock/proxy/api/ProxyService.scala [hunk 7]
```diff
@@ -48,7 +48,8 @@ trait ProxyService extends LazyLogging {
             onComplete(areCredentialsActive(s3Request.credential)) {
               case Success(Some(userSTS: User)) =>
                 logger.debug(s"Credentials active for request, user retrieved: $userSTS")
-                processRequestForValidUser(clientIPAddress, headerIPs, httpRequest, s3Request, userSTS)
+                val fullS3Request = s3Request.copy(clientIPAddress = clientIPAddress, headerIPs = headerIPs)
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/airlock/proxy/api/ProxyService.scala]: yes - good point - moved

File: src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala (hunk 3 above)
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala]: done

**Pull request merged**